File: typescript/samples/FileVerifier/file-verifier.ts
import { Topic, type Context } from "../../src/types";
import type {
  FileUploadedMessage,
  FileVerifiedMessage,
} from "../../src/messages";

export const onMessage = async (
  message: FileUploadedMessage,
  { logger, database, fileStorage, emit }: Context<FileVerifiedMessage>,
) => {
  logger.info(`Processing file verification for fileId: ${message.fileId}`);

  const imageExists = await database.withConnection(async (conn) => {
    const result = await conn`
      SELECT id FROM images WHERE id = ${message.fileId}
    `;
    return result.length > 0;
  });

  if (!imageExists) {
    logger.error(`Image with id ${message.fileId} not found in database`);
    throw new Error(`Image not found in database: ${message.fileId}`);
  }

  logger.debug(`Image ${message.fileId} found in database`);

  if (!fileStorage) {
    throw new Error("File storage is not configured");
  }

  try {
    const stream = await fileStorage.openRead({
      container: message.bucketName,
      key: message.objectKey,
    });
    await stream.cancel();
    logger.debug(
      `Image ${message.fileId} exists in storage at ${message.bucketName}/${message.objectKey}`,
    );
  } catch (error) {
    logger.error(
      `Image ${message.fileId} not found in storage: ${message.bucketName}/${message.objectKey}`,
    );
    throw new Error(`Image not found in storage: ${message.objectKey}`);
  }

  await database.withConnection(async (conn) => {
    await conn`
      UPDATE images 
      SET verified = true, verified_at = NOW()
      WHERE id = ${message.fileId}
    `;
  });

  logger.info(`Image ${message.fileId} marked as verified in database`);

  const verifiedMessage: FileVerifiedMessage = {
    fileId: message.fileId,
    fileName: message.fileName,
    contentType: message.contentType,
    bucketName: message.bucketName,
    objectKey: message.objectKey,
    sizeBytes: message.sizeBytes,
    uploadedAt: message.uploadedAt,
  };

  await emit(verifiedMessage, Topic.FileVerified);
  logger.info(`Emitted FileVerified message for ${message.fileId}`);
};

File: dotnet/samples/FileVerifier/FileVerifier.cs
using FileVerifier.Generated;
using TaskRunner.Abstractions;
using AbstractTopic = TaskRunner.Abstractions.Topic;

namespace FileVerifier;

public sealed class FileVerifier : IMessageHandler<FileUploadedMessage, FileVerifiedMessage>
{
    public async Task HandleAsync(FileUploadedMessage message, ITaskRunnerContext<FileVerifiedMessage> context, CancellationToken ct)
    {
        var logger = context.Logger;

        logger.LogInformation("Processing file verification for fileId: {FileId}", message.FileId);

        var imageExists = await context.Database.WithConnectionAsync(async (conn, token) =>
        {
            await using var cmd = conn.CreateCommand();
            cmd.CommandText = "SELECT COUNT(*) FROM images WHERE id = @Id";

            var idParam = cmd.CreateParameter();
            idParam.ParameterName = "@Id";
            idParam.Value = message.FileId;
            cmd.Parameters.Add(idParam);

            var result = await cmd.ExecuteScalarAsync(token);
            return Convert.ToInt32(result) > 0;
        }, ct);

        if (!imageExists)
        {
            logger.LogError("Image with id {FileId} not found in database", message.FileId);
            throw new InvalidOperationException($"Image not found in database: {message.FileId}");
        }

        logger.LogDebug("Image {FileId} found in database", message.FileId);

        if (context.FileStorage is null)
        {
            throw new InvalidOperationException("File storage is not configured");
        }

        try
        {
            var storageId = new StorageObjectId(message.BucketName, message.ObjectKey);
            await using var stream = await context.FileStorage.OpenReadAsync(storageId, ct);
            logger.LogDebug("Image {FileId} exists in storage at {Bucket}/{Key}",
                message.FileId, message.BucketName, message.ObjectKey);
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Image {FileId} not found in storage: {Bucket}/{Key}",
                message.FileId, message.BucketName, message.ObjectKey);
            throw new InvalidOperationException($"Image not found in storage: {message.ObjectKey}", ex);
        }

        await context.Database.WithConnectionAsync(async (conn, token) =>
        {
            await using var cmd = conn.CreateCommand();
            cmd.CommandText = "UPDATE images SET verified = true, verified_at = @VerifiedAt WHERE id = @Id";

            var idParam = cmd.CreateParameter();
            idParam.ParameterName = "@Id";
            idParam.Value = message.FileId;
            cmd.Parameters.Add(idParam);

            var verifiedAtParam = cmd.CreateParameter();
            verifiedAtParam.ParameterName = "@VerifiedAt";
            verifiedAtParam.Value = DateTimeOffset.UtcNow;
            cmd.Parameters.Add(verifiedAtParam);

            await cmd.ExecuteNonQueryAsync(token);
        }, ct);

        logger.LogInformation("Image {FileId} marked as verified in database", message.FileId);

        var verifiedMessage = new FileVerifiedMessage(
            fileId: message.FileId,
            fileName: message.FileName,
            contentType: message.ContentType,
            bucketName: message.BucketName,
            objectKey: message.ObjectKey,
            sizeBytes: message.SizeBytes,
            uploadedAt: message.UploadedAt
        );

        await context.EmitAsync(verifiedMessage, AbstractTopic.FileVerified, ct);
        logger.LogInformation("Emitted FileVerified message for {FileId}", message.FileId);
    }
}
